File: packages/guards/src/pii-de.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

type PiiDeEntity = 'steuer-id' | 'personalausweis' | 'reisepass' | 'iban-de' | 'phone-de';

interface PiiDeOptions {
  entities: PiiDeEntity[];
  action: 'block' | 'warn' | 'mask';
}

interface PiiDeMatch { type: PiiDeEntity; value: string; start: number; end: number; }

const PATTERNS: Record<PiiDeEntity, RegExp> = {
  'steuer-id': /\b\d{2}\s?\d{3}\s?\d{3}\s?\d{3}\b/g,
  personalausweis: /\b[CFGHJKLMNPRTVWXYZ]\d{8}\b/g,
  reisepass: /\b[CFGHJKLMNPRTVWXYZ]\d{8}\b/g,
  'iban-de': /\bDE\d{2}\s?\d{4}\s?\d{4}\s?\d{4}\s?\d{4}\s?\d{2}\b/g,
  'phone-de': /(?:\+49|0049)\s?\d{2,5}[-/\s]?\d{3,9}/g,
};

const MASK_LABELS: Record<PiiDeEntity, string> = {
  'steuer-id': '[Steuer-ID]',
  personalausweis: '[Personalausweis]',
  reisepass: '[Reisepass]',
  'iban-de': '[IBAN]',
  'phone-de': '[Telefonnummer]',
};

function detect(text: string, entities: PiiDeEntity[]): PiiDeMatch[] {
  const matches: PiiDeMatch[] = [];
  for (const entity of entities) {
    const re = new RegExp(PATTERNS[entity].source, 'g');
    let m: RegExpExecArray | null;
    while ((m = re.exec(text)) !== null) {
      matches.push({ type: entity, value: m[0], start: m.index, end: m.index + m[0].length });
    }
  }
  return matches.sort((a, b) => b.start - a.start);
}

function mask(text: string, matches: PiiDeMatch[]): string {
  let r = text;
  for (const m of matches) { r = r.slice(0, m.start) + MASK_LABELS[m.type] + r.slice(m.end); }
  return r;
}

export function piiDe(options: PiiDeOptions): Guard {
  return {
    name: 'pii-de', version: '0.1.0',
    description: 'German PII detection (Steuer-ID, Personalausweis, IBAN, phone)',
    category: 'privacy', supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matches = detect(text, options.entities);
      const triggered = matches.length > 0;
      if (!triggered) return { guardName: 'pii-de', passed: true, action: 'allow', latencyMs: Math.round(performance.now() - start) };
      if (options.action === 'mask') {
        return { guardName: 'pii-de', passed: true, action: 'override', overrideText: mask(text, matches), latencyMs: Math.round(performance.now() - start), details: { detected: matches.map(({ type, value }) => ({ type, value })) } };
      }
      return { guardName: 'pii-de', passed: false, action: options.action, message: `German PII detected: ${[...new Set(matches.map((m) => m.type))].join(', ')}`, latencyMs: Math.round(performance.now() - start), details: { detected: matches.map(({ type, value }) => ({ type, value })) } };
    },
  };
}

File: python/open_guardrail/guards/pii_de.py
"""German PII detection and masking guard."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PATTERNS: dict[str, re.Pattern[str]] = {
    "personalausweis": re.compile(r"\b[CFGHJKLMNPRTVWXYZ0-9]{9}\d\b"),
    "steuer-id": re.compile(r"\b\d{11}\b"),
    "sozialversicherung": re.compile(r"\b\d{2}\s?\d{6}\s?[A-Z]\s?\d{3}\b"),
    "iban-de": re.compile(r"\bDE\d{2}\s?\d{4}\s?\d{4}\s?\d{4}\s?\d{4}\s?\d{2}\b", re.I),
    "phone-de": re.compile(r"\b(?:\+49|0049|0)\s?(?:\d[\s\-]?){9,11}\b"),
}

_MASK_LABELS: dict[str, str] = {
    "personalausweis": "[PERSONALAUSWEIS-NR]",
    "steuer-id": "[STEUER-ID]",
    "sozialversicherung": "[SOZIALVERSICHERUNGSNR]",
    "iban-de": "[IBAN-DE]",
    "phone-de": "[TELEFON-DE]",
}


def _detect(text: str, entities: list[str]) -> list[dict]:
    matches: list[dict] = []
    for entity in entities:
        pat = _PATTERNS.get(entity)
        if not pat:
            continue
        for m in pat.finditer(text):
            matches.append({"type": entity, "value": m.group(), "start": m.start(), "end": m.end()})
    matches.sort(key=lambda x: x["start"], reverse=True)
    return matches


def _mask(text: str, matches: list[dict]) -> str:
    result = text
    for m in matches:
        result = result[: m["start"]] + _MASK_LABELS.get(m["type"], "[PII]") + result[m["end"] :]
    return result


class _PiiDe:
    def __init__(self, *, action: str = "block", entities: Optional[List[str]] = None) -> None:
        self.name = "pii-de"
        self.action = action
        self._entities = entities or list(_PATTERNS.keys())

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matches = _detect(text, self._entities)
        elapsed = (time.perf_counter() - start) * 1000
        if not matches:
            return GuardResult(guard_name="pii-de", passed=True, action="allow", latency_ms=round(elapsed, 2))
        detected = [{"type": m["type"], "value": m["value"]} for m in matches]
        if self.action == "mask":
            return GuardResult(guard_name="pii-de", passed=True, action="override", override_text=_mask(text, matches), latency_ms=round(elapsed, 2), details={"detected": detected})
        return GuardResult(guard_name="pii-de", passed=False, action=self.action, message="German PII detected", latency_ms=round(elapsed, 2), details={"detected": detected})


def pii_de(*, action: str = "block", entities: Optional[List[str]] = None) -> _PiiDe:
    return _PiiDe(action=action, entities=entities)
